Adding example for cert-manager letsencrypt for Azure Servicemesh (#4511)
Azure/AKS: examples/azure-service-mesh/security/cert-manager-letsencrypt/readme.md

> # Cert-Manager Let's Encrypt Integration with Azure Service Mesh (ASM)

Reviewer: we need to use Istio based service mesh add-on for AKS instead of Azure Service Mesh and ASM .

> This document contains instructions on how to integrate Azure Servicemesh with Cert-Manager and obtain letsencrypt certificates. 

Reviewer: ```suggestion
This document contains instructions on how to integrate Istio based service mesh add-on for AKS with cert-manager and obtain letsencrypt certificates for setting up secure ingress gateways. 
```

## Objectives
* Deploy bookinfo demo app, expose a secure HTTPS service using simple TLS.
* Demonstrate HTTPS connections for Azure Service Mesh workloads using Cert-Manager and letsencrypt as the certificate authority.

## Before you begin
* Install Install Azure Servicemesh.
* Enable external ingressgateway.
* Enable sidecar injection on the default namespace. 

```shell
az aks mesh enable -g <rg-name> -n <cluster-name>
az aks mesh enable-ingress-gateway -g <rg-name> -n <cluster-name> --ingress-gateway-type external
revision=$(az aks show --resource-group <rg-name> --name <cluster-name> --query 'serviceMeshProfile.istio.revisions[0]' -o tsv)
kubectl label namespace default istio.io/rev=$revision
```

## Steps
### 1. Setup DNS record

> Setup a DNS record for the external-ip address for the external ingressgateway service. In this example, I set up 4.153.8.39 with test.dev.azureservicemesh.io

Author: I'd really defer this to the user, we've provided ample information to them for the dns record setup.

```shell
kubectl get svc -n aks-istio-ingress
```

```console
NAME                                TYPE           CLUSTER-IP   EXTERNAL-IP   PORT(S)                                      AGE
aks-istio-ingressgateway-external   LoadBalancer   10.0.59.31   4.153.8.39    15021:30786/TCP,80:30626/TCP,443:30236/TCP   8m44s
```

> Verify/wait until `dig +short A test.dev.azureservicemesh.io` returns the configured IP address that is `4.153.8.39` in this example.

Author: I'd really defer this to the user, we've provided ample information to them for the dns record setup.

Create a ConfigMap with the name `istio-shared-configmap-<asm-revision>` in the `aks-istio-system` namespace to set `ingressService` and `ingressSelector`. For example, if your cluster is running asm-1-21 revision of mesh, then the ConfigMap needs to be named as istio-shared-configmap-asm-1-21. Mesh configuration has to be provided within the data section under mesh.

```shell
kubectl apply -f configmap.yaml
```

### 5. Install cert-manager
```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.15.2/cert-manager.yaml
```
this installs cert-manager in its own namespace `cert-manager`

#### Validate cert-manager pods
```shell
kubectl get po -n cert-manager
```
```console
NAME                                     READY   STATUS    RESTARTS   AGE
cert-manager-6fd987499c-xmf44            1/1     Running   0          72s
cert-manager-cainjector-5b94bd6f-jw644   1/1     Running   0          72s
cert-manager-webhook-575479ff47-d87pf    1/1     Running   0          72s
```

### 6. Setup cluster-issuer and Certificate resources
Set your email address in `cluster-issuer.yaml` that you'd like to register with ACME server.
```shell

> kubectl apply -f cluster-issuer.yaml

Reviewer: need to call out allowed vs supported distinction - https://learn.microsoft.com/en-us/azure/aks/istio-support-policy#allowed-supported-and-blocked-customizations

# Cert-Manager Let's Encrypt Integration with Istio-based service mesh add-on

This document contains instructions on how to integrate Istio-based service mesh add-on for AKS with cert-manager and obtain letsencrypt certificates for setting up secure ingress gateways. 

## Objectives
* Deploy bookinfo demo app, expose a secure HTTPS service using simple TLS.

> * Demonstrate HTTPS connections for Azure Service Mesh workloads using Cert-Manager and letsencrypt as the certificate authority.

Reviewer: nit: cert-manager and Let's Encrypt everywhere for consistency

Author: updated

## Before you begin

> * Install Install Azure Servicemesh.

Reviewer: ```suggestion
* Install Istio-based service mesh add-on on your cluster
```

Reviewer: and we should share a link to this page - https://learn.microsoft.com/azure/aks/istio-deploy-addon

> * Enable external ingressgateway.

Reviewer: Same as above, we need to add a link to product docs

Author: added appropriate links

* Enable sidecar injection on the default namespace. 

```shell
az aks mesh enable -g <rg-name> -n <cluster-name>
az aks mesh enable-ingress-gateway -g <rg-name> -n <cluster-name> --ingress-gateway-type external
revision=$(az aks show --resource-group <rg-name> --name <cluster-name> --query 'serviceMeshProfile.istio.revisions[0]' -o tsv)
kubectl label namespace default istio.io/rev=$revision

> ```

Reviewer: Ah I see you have commands listed below the bullet points. the section above said before you begin. would recommend interleaving these commands in between those points to avoid confusion.

## Steps
### 1. Setup DNS record

> Setup a DNS record for the external-ip address for the external ingressgateway service. In this example, I set up 4.153.8.39 with test.dev.azureservicemesh.io

Reviewer: Set up DNS record where? Would be good to provide some pointers on that

Author: this is hard as this is going to be different for each cloud provider. I've just mentioned `with your cloud provider`

### Validate http request to productpage service

Send an HTTP request to access the productpage service
```shell
curl -vs http://test.dev.azureservicemesh.io/productpage | grep -o "<title>.*</title>"
```

this should print `<title>Simple Bookstore App</title>`

### 4. Create the shared configmap
Create a ConfigMap with the name `istio-shared-configmap-<asm-revision>` in the `aks-istio-system` namespace to set `ingressService` and `ingressSelector`. For example, if your cluster is running asm-1-21 revision of mesh, then the ConfigMap needs to be named as istio-shared-configmap-asm-1-21. Mesh configuration has to be provided within the data section under mesh.

```shell
kubectl apply -f configmap.yaml
```

> [!Note]  
> Kubernetes ingress for Istio-based service mesh is an `allowed` feature. More details on configuration options [here](https://learn.microsoft.com/en-us/azure/aks/istio-support-policy#allowed-supported-and-blocked-customizations)

> ### 5. Install cert-manager

Reviewer: Add a note (maybe here and at the start of this page) that this is an example for self-managed cert-manager installation. And that cert-manager itself is not covered by support boundary and that its lifecycle is self-managed

```shell
kubectl apply -f cluster-issuer.yaml
kubectl apply -f certificate.yaml
```
This should create k8s secret `bookinfo-certs` in `aks-istio-ingress` ns as requested by the certificate resource created above.

```shell
kubectl get secret -n aks-istio-ingress
```

```console
NAME                                                                TYPE                 DATA   AGE
bookinfo-certs                                                      kubernetes.io/tls    2      70s
sh.helm.release.v1.asm-igx-aks-istio-ingressgateway-external.v112   helm.sh/release.v1   1      3m19s
sh.helm.release.v1.asm-igx-aks-istio-ingressgateway-external.v113   helm.sh/release.v1   1      80s
```

### Validate https request to productpage service
Verify that the productpage can be accessed via the HTTPS endpoint

```shell
curl -v https://test.dev.azureservicemesh.io/productpage | grep -o "<title>.*</title>"
```
this should print `<title>Simple Bookstore App</title>`

## Troubleshooting
1. Verify that the DNS record has propagated
```dig +short A test.dev.azureservicemesh.io``` should return the configured ip address ```4.153.8.39``` in the example.

> 2. check certificate resource readiness

Reviewer: formatting - capitalize first letter

```
```console
NAME                         READY   SECRET           AGE
bookinfo-letsencrypt-certs   False   bookinfo-certs   2m5s
```
In case certificate resource is not ready as above, check the resources provisioned by cert-manager in order to solve the ACME challenge. Below is the chain of resources created by cert-manager.

```shell
kubectl describe certificate bookinfo-letsencrypt-certs -n aks-istio-ingress
```
```console
Name:         bookinfo-letsencrypt-certs
Namespace:    aks-istio-ingress
Labels:       <none>
Annotations:  <none>
API Version:  cert-manager.io/v1
Kind:         Certificate
...
Events:
  Type    Reason     Age   From                                       Message
  ----    ------     ----  ----                                       -------
  Normal  Issuing    14m   cert-manager-certificates-trigger          Issuing certificate as Secret does not exist
  Normal  Generated  14m   cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "bookinfo-letsencrypt-certs-hpjst"
  Normal  Requested  14m   cert-manager-certificates-request-manager  Created new CertificateRequest resource "bookinfo-letsencrypt-certs-1"
```

> 3. check certificaterequest resource.

Reviewer: what does the user need to check inside this resource?

Author: Just wanted to let the users know that these are the chain of resources provisioned by cert-manager and they should troubleshoot these in case something doesn't work for them. Added the header to check for readiness of these resources.

  ----    ------              ----   ----                                                -------
  Normal  WaitingForApproval  6m57s  cert-manager-certificaterequests-issuer-selfsigned  Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  6m57s  cert-manager-certificaterequests-issuer-vault       Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  6m57s  cert-manager-certificaterequests-issuer-venafi      Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  6m57s  cert-manager-certificaterequests-issuer-acme        Not signing CertificateRequest until it is Approved
  Normal  WaitingForApproval  6m57s  cert-manager-certificaterequests-issuer-ca          Not signing CertificateRequest until it is Approved
  Normal  cert-manager.io     6m57s  cert-manager-certificaterequests-approver           Certificate request has been approved by cert-manager.io
  Normal  OrderCreated        6m57s  cert-manager-certificaterequests-issuer-acme        Created Order resource aks-istio-ingress/bookinfo-letsencrypt-certs-1-507153367
  ```

> 4. check order resource.

Reviewer: what does the user need to check inside this resource?

Author: Just wanted to let the users know that these are the chain of resources provisioned by cert-manager and they should troubleshoot these in case something doesn't work for them. Added the header to check for readiness of these resources.

```shell
kubectl describe order bookinfo-letsencrypt-certs-1-507153367 -n aks-istio-ingress
```
```console
Name:         bookinfo-letsencrypt-certs-1-507153367
Namespace:    aks-istio-ingress
Labels:       <none>
Annotations:  cert-manager.io/certificate-name: bookinfo-letsencrypt-certs
              cert-manager.io/certificate-revision: 1
              cert-manager.io/private-key-secret-name: bookinfo-letsencrypt-certs-hpjst
...
Events:
  Type    Reason   Age   From                 Message
  ----    ------   ----  ----                 -------
  Normal  Created  11m   cert-manager-orders  Created Challenge resource "bookinfo-letsencrypt-certs-1-507153367-692923294" for domain "test.dev.azureservicemesh.io"
```

> 5. check challenge resource.

Reviewer: what does the user need to check inside this resource?

Author: Just wanted to let the users know that these are the chain of resources provisioned by cert-manager and they should troubleshoot these in case something doesn't work for them. Added the header to check for readiness of these resources.

This document contains instructions on how to integrate Istio-based service mesh add-on for AKS with cert-manager and obtain let's encrypt certificates for setting up secure ingress gateways. 

## Objectives
* Deploy bookinfo demo app, expose a secure HTTPS service using simple TLS.
* Demonstrate HTTPS connections for Azure Service Mesh workloads using cert-manager and let's encrypt as the certificate authority.

## Before you begin
* [Install](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-addon#install-istio-add-on) Istio-based service mesh add-on on your cluster.
```shell
az aks mesh enable -g <rg-name> -n <cluster-name>
```
* [Enable external ingressgateway](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-ingress#enable-external-ingress-gateway)
```shell
az aks mesh enable-ingress-gateway -g <rg-name> -n <cluster-name> --ingress-gateway-type external
```
* [Enable sidecar injection](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-addon#enable-sidecar-injection) on the default namespace. 
```shell
revision=$(az aks show --resource-group <rg-name> --name <cluster-name> --query 'serviceMeshProfile.istio.revisions[0]' -o tsv)

> kubectl label namespace default istio.io/rev=$revision

Reviewer 2: place holders and env vars mixed. Where is `$revision` defined?

Author: line 20 defines revision variable.

Reviewer 3: Should we use env vars for resource group and cluster-name so it's easier for folks to copy paste?

Author: I had placeholders as these are not extensively used throughout the walkthrough. 
Only within `Before you begin` section.

```

Confirm several deployments and services are created on your cluster. For example:
```console
service/details created
serviceaccount/bookinfo-details created
deployment.apps/details-v1 created
service/ratings created
serviceaccount/bookinfo-ratings created
deployment.apps/ratings-v1 created
service/reviews created
serviceaccount/bookinfo-reviews created
deployment.apps/reviews-v1 created
deployment.apps/reviews-v2 created
deployment.apps/reviews-v3 created
service/productpage created
serviceaccount/bookinfo-productpage created
deployment.apps/productpage-v1 created
```
```shell
kubectl get pods
```
Ensure each pod has 2/2 containers in the `Ready` state
```console
NAME                              READY   STATUS    RESTARTS   AGE
details-v1-558b8b4b76-2llld       2/2     Running   0          2m41s
productpage-v1-6987489c74-lpkgl   2/2     Running   0          2m40s
ratings-v1-7dc98c7588-vzftc       2/2     Running   0          2m41s
reviews-v1-7f99cc4496-gdxfn       2/2     Running   0          2m41s
reviews-v2-7d79d5bd5d-8zzqd       2/2     Running   0          2m41s
reviews-v3-7dbcdcbc56-m8dph       2/2     Running   0          2m41s
```

> ### 3. Configure ingress gateway and virtual service

Reviewer 2: Remind users to update host name in both files.

# Cert-Manager Let's Encrypt Integration with Istio-based service mesh add-on

> This document contains instructions on how to integrate Istio-based service mesh add-on for AKS with cert-manager and obtain let's encrypt certificates for setting up secure ingress gateways. 

Reviewer 3: Should we call out support policy about third-party integrations with cert-manager and LetsEncrypt, and K8s ingress, being allowed but unsupported? @shashankbarsin @biefy

Reviewer 3: Or is that already implied by having it in the examples dir?

Author: this has already been been addressed as part of a note under `Create the shared configmap` section.
@nshankar13

Reviewer 3: I saw - left a comment there.

## Objectives
* Deploy bookinfo demo app, expose a secure HTTPS service using simple TLS.
* Demonstrate HTTPS connections for Azure Service Mesh workloads using cert-manager and let's encrypt as the certificate authority.

## Before you begin
* [Install](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-addon#install-istio-add-on) Istio-based service mesh add-on on your cluster.
```shell
az aks mesh enable -g <rg-name> -n <cluster-name>
```
* [Enable external ingressgateway](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-ingress#enable-external-ingress-gateway)
```shell

> az aks mesh enable-ingress-gateway -g <rg-name> -n <cluster-name> --ingress-gateway-type external

Reviewer 3: Would internal ingress gateway also work or does it have to be an external IP? If so I think we should add a note here and a comment in configmap.yaml that if they're using internal, then they'd need to update the ingressService and ingressSelector accordingly.

Author: `ingressSelector` and `ingressService` needs to be set to external for ACME challenge path resolved.

Reviewer 2: Internal would not work because http challenge won't be served properly.

```
* [Enable sidecar injection](https://learn.microsoft.com/en-us/azure/aks/istio-deploy-addon#enable-sidecar-injection) on the default namespace. 
```shell
revision=$(az aks show --resource-group <rg-name> --name <cluster-name> --query 'serviceMeshProfile.istio.revisions[0]' -o tsv)
kubectl label namespace default istio.io/rev=$revision
```

## Steps
### 1. Setup DNS record

> Set up a DNS record for the `EXTERNAL-IP` address of the external ingressgateway service with your cloud provider. In this example, we are setting up the DNS record for `4.153.8.39` with `test.dev.azureservicemesh.io`.

Reviewer 3: Let's link to relevant Azure docs for setting the DNS record?

Reviewer 2: The customer may or may not use Azure DNS.  I think it is fair to assume readers know how to add a DNS record.

```shell
kubectl apply -f gateway.yaml
kubectl apply -f virtualservice.yaml
```

> [!NOTE]  
> In the gateway definition, `credentialName` must match the `secretName` in the `certificate` resource which will be created later in the example and selector must refer to the external ingress gateway by its label, in which the key of the label is `istio` and the value is `aks-istio-ingressgateway-external`.

### Validate http request to productpage service

Send an HTTP request to access the productpage service
```shell
curl -vs http://test.dev.azureservicemesh.io/productpage | grep -o "<title>.*</title>"
```

this should print `<title>Simple Bookstore App</title>`

### 4. Create the shared configmap
Create a ConfigMap with the name `istio-shared-configmap-<asm-revision>` in the `aks-istio-system` namespace to set `ingressService` and `ingressSelector`. For example, if your cluster is running asm-1-21 revision of mesh, then the ConfigMap needs to be named as istio-shared-configmap-asm-1-21. Mesh configuration has to be provided within the data section under mesh.

```shell
kubectl apply -f configmap.yaml
```

> [!Note]  

> > Kubernetes ingress for Istio-based service mesh is an `allowed` feature. More details on configuration options [here](https://learn.microsoft.com/en-us/azure/aks/istio-support-policy#allowed-supported-and-blocked-customizations)  

Reviewer 3: Oh I see this - maybe we want to move this call out to the top of the readme?

Author: moved.

productpage-v1-6987489c74-lpkgl   2/2     Running   0          2m40s
ratings-v1-7dc98c7588-vzftc       2/2     Running   0          2m41s
reviews-v1-7f99cc4496-gdxfn       2/2     Running   0          2m41s
reviews-v2-7d79d5bd5d-8zzqd       2/2     Running   0          2m41s
reviews-v3-7dbcdcbc56-m8dph       2/2     Running   0          2m41s
```
### 3. Configure ingress gateway and virtual service
Before deploying the `virtualservice` and `gateway` resources, make sure to update the host name to match your own DNS name.
```shell
kubectl apply -f gateway.yaml
kubectl apply -f virtualservice.yaml
```

> [!NOTE]  
> In the gateway definition, `credentialName` must match the `secretName` in the `certificate` resource which will be created later in the example and selector must refer to the external ingress gateway by its label, in which the key of the label is `istio` and the value is `aks-istio-ingressgateway-external`.

### Validate http request to productpage service

Send an HTTP request to access the productpage service
```shell
curl -vs http://test.dev.azureservicemesh.io/productpage | grep -o "<title>.*</title>"
```

> this should print `<title>Simple Bookstore App</title>`

Reviewer 4: It prints also the headers because of `curl -v`. You can remove the `-v`

### 4. Create the shared configmap
Create a ConfigMap with the name `istio-shared-configmap-<asm-revision>` in the `aks-istio-system` namespace to set `ingressService` and `ingressSelector`. For example, if your cluster is running asm-1-21 revision of mesh, then the ConfigMap needs to be named as istio-shared-configmap-asm-1-21. Mesh configuration has to be provided within the data section under mesh.

```shell

> kubectl apply -f configmap.yaml

Reviewer 4: You can make the process easier giving a command that replaces the value.
```suggestion
cat configmap.yaml | sed -e "s/<asm-revision>/${revision}/" |  kubectl apply -f -
```

```

### 5. Install cert-manager
```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.15.2/cert-manager.yaml
```
this installs cert-manager in its own namespace `cert-manager`

#### Validate cert-manager pods
```shell
kubectl get po -n cert-manager
```
```console
NAME                                     READY   STATUS    RESTARTS   AGE
cert-manager-6fd987499c-xmf44            1/1     Running   0          72s
cert-manager-cainjector-5b94bd6f-jw644   1/1     Running   0          72s
cert-manager-webhook-575479ff47-d87pf    1/1     Running   0          72s
```

### 6. Setup cluster-issuer and Certificate resources
Set your email address in `cluster-issuer.yaml` that you'd like to register with ACME server.
```shell

> kubectl apply -f cluster-issuer.yaml

Reviewer 4: Why the `IngressClass` is in this file ? Do you want to move it to a dedicated file to show the customer that this is also being created ?

Also, to avoid the customer editing your template files:

```suggestion
EMAIL=email
cat cluster-issuer.yaml | sed  -e "s/your-email-addr@domain.com/${EMAIL}/" | kubectl apply -f -
```

kubectl apply -f certificate.yaml
```

> This should create k8s secret `bookinfo-certs` in `aks-istio-ingress` ns as requested by the certificate resource created above.

Reviewer 4: ```suggestion
This should create k8s secret `bookinfo-certs` in `aks-istio-ingress` namespace as requested by the certificate resource created above.
```

```shell
kubectl get secret -n aks-istio-ingress
```

```console
NAME                                                                TYPE                 DATA   AGE
bookinfo-certs                                                      kubernetes.io/tls    2      70s
sh.helm.release.v1.asm-igx-aks-istio-ingressgateway-external.v112   helm.sh/release.v1   1      3m19s
sh.helm.release.v1.asm-igx-aks-istio-ingressgateway-external.v113   helm.sh/release.v1   1      80s
```

### Validate https request to productpage service
Verify that the productpage can be accessed via the HTTPS endpoint

```shell
curl -v https://test.dev.azureservicemesh.io/productpage | grep -o "<title>.*</title>"
```

> this should print `<title>Simple Bookstore App</title>`

Reviewer 4: Again drop the `-v` from curl
